File: 백준/Silver/2529. 부등호/부등호.cc
#include<bits/stdc++.h>
using namespace std;
int k;
char c[10];
string mx="0",mn="99999999999";
bool visited[10];
bool oper(int a,int b,char cc)
{
	if(cc=='>') return a>b;
	else if(cc=='<') return a<b;
}
void dfs(int idx,string s)
{
	if(idx==k+1)
	{
		if(mx<s) mx=s;
		if(mn>s) mn=s;
		return;
	}
	for(int i=0;i<=9;i++)
	{
		if(visited[i]) continue;
		if(idx>=1 && !oper(s[idx-1]-'0',i,c[idx-1])) continue;
		visited[i]=true;
		dfs(idx+1,s+to_string(i));
		visited[i]=false;
	}
}
int main()
{
	cin>>k;
	for(int i=0;i<k;i++)
	{
		cin>>c[i];
	}
	dfs(0,"");
	cout<<mx<<"\n"<<mn;
}

File: 백준/Silver/2529. 부등호/부등호.py
mx="0"
mn="9999999999"

def oper(ch,a,b):
    if ch=='<':
        return a<b
    elif ch=='>':
        return a>b

def dfs(idx,num):
    global mx,mn
    if idx==k+1:
        if mx<num:
            mx=num
        if mn>num:
            mn=num
        return

    for i in range(10):
        if visited[i]:
            continue
        if idx>=1 and not oper(op[idx-1],int(num[idx-1]),i):
            continue
        visited[i]=True
        dfs(idx+1,num+str(i))
        visited[i]=False

k=int(input().strip())
op=list(input().split())
visited=[False]*10
dfs(0,"") # index, num(str)
print(mx)
print(mn)